pkg/cli/nodelocal.go

```diff
@@ -0,0 +1,109 @@
+// Copyright 2019 The Cockroach Authors.
+//
+// Use of this software is governed by the Business Source License
+// included in the file licenses/BSL.txt.
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0, included in the file
+// licenses/APL.txt.
+
+package cli
+
+import (
+	"database/sql/driver"
+	"fmt"
+	"io"
+	"os"
+
+	"github.com/cockroachdb/cockroach/pkg/sql"
+	"github.com/pkg/errors"
+	"github.com/spf13/cobra"
+)
+
+const (
+	chunkSize = 4 * 1024
+)
+
+var nodeLocalUploadCmd = &cobra.Command{
+	Use:   "upload <source> <destination>",
+	Short: "Upload file from source to destination",
+	Long: `
+Uploads a file to a gateway node's local file system using a SQL connection.
+`,
+	Args: cobra.MinimumNArgs(2),
+	RunE: maybeShoutError(runUpload),
+}
+
+//var verUploadFile = version.MustParse("")
+
+func runUpload(cmd *cobra.Command, args []string) error {
+	conn, err := makeSQLClient("cockroach nodelocal", useSystemDb)
+	if err != nil {
+		return err
+	}
+	defer conn.Close()
+	// TODO: which version should we be checking for?
+	//if err := conn.requireServerVersion(verUploadFile); err != nil {
+	//	return err
+	//}
+
+	source := args[0]
+	destination := args[1]
+	reader, err := openSourceFile(source)
+	if err != nil {
+		return err
+	}
+	return uploadFile(conn, reader, destination)
+}
+
+func openSourceFile(source string) (io.ReadCloser, error) {
+	stat, err := os.Stat(source)
+	if err != nil {
+		return nil, errors.Wrapf(err, "unable to get source file stats for %s", source)
+	}
+	if stat.IsDir() {
+		return nil, fmt.Errorf("source file %s is a directory, not a file", source)
+	}
+	return os.Open(source)
+}
+
+func uploadFile(conn *sqlConn, reader io.ReadCloser, destination string) error {
+	return conn.ExecTxn(func(cn *sqlConn) error {
+		stmt, err := cn.conn.Prepare(sql.CopyInFileStmt(destination, "crdb_internal", "file_upload"))
+		if err != nil {
+			return err
+		}
+		defer stmt.Close()
+		for {
+			send := make([]byte, chunkSize)
+			n, err := reader.Read(send)
+			if err == io.EOF {
```

Comment:
n can be > 0 even with an EOF, so you still need to send the number of bytes read. i'm surprised the tests passed.

Ah, it's because the API for io.Reader is that n can be > 0 even with EOF. But os.File.Read specifies 0, EOF at the end. You have two choices:

1) Teach the code here to always deal with n > 0. You should do this (as per the io.Reader API) only checking for error if n == 0, since that API says any error should be returned during subsequent calls. This means you don't have to do any weird error checking after processing the data, which is nice.
2) Have this function take an os.File, which guarantees that 0, EOF is returned on EOF.

Both of these are fine and you could do whichever you like more. I'd probably do the first since it feels safer and is also more testable since it doesn't depend on a file coming in.

Reply:
Yes, I will do the first. It would cause less confusion if we ever used any other implementation of reader.